cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,673 @@
+                                    meta={
+                                        'variable_name': in_edge.source.identifier,
+                                        'time_lag': in_edge.source.time_lag + lag,
+                                    },
+                                )
+                                extended_graph.add_node(node=lagged_source_node)
+                            else:
+                                lagged_source_node = extended_graph.get_node(lagged_source_identifier)
+                            lagged_edge = Edge(
+                                source=lagged_source_node,
+                                destination=lagged_node,
+                                edge_type=in_edge.get_edge_type(),
+                            )
+
+                            extended_graph.add_edge(edge=lagged_edge)
+
+        return extended_graph
+
+    @staticmethod
+    def get_variable_names_from_node_names(node_names: List[str]):
+        """
+        Return a list of variable names from a list of node names.
+        This is useful for converting a list of node names into a list of variable names.
+        Variables are the elementary (unique) units of a time series causal graph.
+
+        Example:
+            ['X', 'X lag(n=1)', 'Y', 'Z lag(n=2)'] -> ['X', 'Y', 'Z']
+
+        :param node_names: Can be a list of node names or a list of dictionaries with node names and lags.
+        :return: List of variable names.
+        """
+        assert isinstance(node_names, list)
+
+        for node_name in node_names:
+            vname, _ = get_variable_name_and_lag(node_name)
+            if vname not in node_names:
+                node_names.append(vname)
+
+        return sorted(node_names)
+
+    @reset_attributes
+    def add_node(
+        self,
+        /,
+        identifier: Optional[NodeLike] = None,
+        *,
+        variable_type: NodeVariableType = NodeVariableType.UNSPECIFIED,
+        meta: Optional[dict] = None,
+        node: Optional[Node] = None,
+        **kwargs,
+    ) -> Node:
+        """
+        Add a node to the graph. See `cai_causal_graph.causal_graph.CausalGraph.add_node` for more details.
+
+        In addition to the `CausalGraph.add_node` method, this method also populates the metadata of the node with
+        the variable name and the time lag.
+
+        :param identifier: The identifier of the node.
+        :param variable_type: The type of the variable.
+        :param meta: The metadata of the node.
+        :param node: The node to add.
+        :param kwargs: Additional keyword arguments.
+        :return: The added node.
+        """
+        node = super().add_node(identifier, variable_type=variable_type, meta=meta, node=node, **kwargs)
+        # populate the metadata for each node
+        vname, lag = get_variable_name_and_lag(node.identifier)
+        node.meta[self._meta_variable_name] = vname
+        node.meta[self._meta_time_name] = lag
```

Comment:
what if the lag is now bigger than the existing order of the graph? should you change the order of the graph here?

Reply:
I could also change it here. But the decorator has this purpose as well. Once I modify the graph, the order and other attributes will be reset to None. Once the order is invoked, then it's computed again.

Reply:
Gotcha makes sense, thanks!